File: solutions/ruby/sum-of-multiples/2/sum_of_multiples.rb
class SumOfMultiples

  MULTIPLE_OF = -> factors, number do
    factors.any? do |factor| factor.nonzero? and
      number.modulo(factor).zero?
    end
  end

  private

  attr_accessor :factors

  def initialize *factors
    self.factors = factors
  end

  public

  def to upper_bound
    upper_bound.times.sum do |number|
      MULTIPLE_OF.(factors, number) ? number : 0
    end
  end

end

File: solutions/haskell/sum-of-multiples/2/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit =
  sum $ filter ((`any` validFactors) . divides) [1..limit-1]
  where
    validFactors = filter (/= 0) factors
    divides n f  = n `mod` f == 0
